Intel GPU oneDNN upstreaming: Conv primitive integration

ZhiweiYan-96 opened the pull request:
# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117529
* __->__ #117512



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ZhiweiYan-96 committed:
Intel GPU oneDNN upstreaming: Conv primitive integration

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

xiaolil1 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.h:
```diff
@@ -0,0 +1,439 @@
+#include <ATen/record_function.h>
+#include <c10/core/MemoryFormat.h>
+
+#include "Attr.h"
+#include "Utils.h"
```
```suggestion
#include <ATen/native/xpu/detail/Attr.h>
#include <ATen/native/xpu/detail/Utils.h>
```

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leizhenyuan committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.h:
```diff
@@ -0,0 +1,446 @@
+constexpr int src_batch_size_dim = 0;
+constexpr int wgh_dst_channels_dim = 0;
+
+static inline dnnl::memory::dims conv_dst_tz(
```
what does "tz" mean?

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.h:
```diff
@@ -0,0 +1,446 @@
+  }
+}
+
+static inline dnnl::memory::format_tag conv_wgh_fmt(
```
what does "wgh" mean?

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Conv.h:
Can we use full name or consistent and well-known abbrev. for variable names?

ZhiweiYan-96 replied:
Thanks for the suggestion, I would rename alll this variables.

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Deconv.h:
ditto on variable names.

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Deconv.h:
```diff
@@ -0,0 +1,417 @@
+          diff_src, weight, diff_dst, groups, is_channels_last_suggested);
+
+  dnnl::memory::format_tag bia_fmt = dnnl::memory::format_tag::x;
+  auto bia_md = bias_defined
```
```suggestion
  auto bias_md = bias_defined
```

ZhiweiYan-96 replied:
has fixed, thanks

jgong5 commented on aten/src/ATen/native/mkldnn/xpu/detail/Deconv.h:
```diff
@@ -0,0 +1,417 @@
+  // create fwd primitive desc hint
+  dnnl::primitive_attr pattr;
+  pattr.set_scratchpad_mode(dnnl::scratchpad_mode::user);
+  #if ONEDNN_SUPPORT_DETERMINISTIC
```
why do we need the build macro to control determinism? Shouldn't it be a runtime check?

ZhiweiYan-96 replied:
hi, @jgong5, deterministic is valid after oneDNN v3.4, which is newer than the oneDNN version in ideep. The marco here is a compilation check for the `version >= 3.4`.   If oneDNN version requirement is not met, code in the marco would fail during compilation.

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP]Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

malfet commented on aten/src/ATen/native/mkldnn/xpu/detail/Utils.cpp:
```diff
@@ -349,4 +349,42 @@ bool binary_valid(
 }
 
+bool is_channels_last(at::MemoryFormat fmt){
+  return (at::MemoryFormat::ChannelsLast == fmt) || (at::MemoryFormat::ChannelsLast3d == fmt);
+}
+
+bool is_smf_channels_last(const Tensor& t){
+  return is_channels_last(t.suggest_memory_format());
+}
+
+bool use_channels_last_for_conv(
+    const at::Tensor& src,
+    const at::Tensor& weight,
+    bool is_transpose){
+
+  if (!src.defined() || src.is_sparse()) {
+    // suggest channels_first
+    return false;
+  }
+
+  auto suggest_channels_last_format =
+      (is_smf_channels_last(src) || is_smf_channels_last(weight));
+  if (suggest_channels_last_format) {
+    // suggest channels_last
+    return true;
+  }
+
+  return false;
+}
+
+std::vector<int64_t> compatible_groups_deconv_strides(
+    const at::Tensor& weight,
+    dnnl::memory::dims group_size) {
+  std::vector<int64_t> strides = weight.strides().vec();
+  strides[0] = weight.strides()[1];
+  strides[1] = weight.strides()[0];
+  strides.insert(strides.begin(), group_size[2] * weight.strides()[0]);
+  return strides;
+}
+
```
Those functions does not seep to be XPU specific (and I would be surprised if we don't have something like that for CPU already)
Can they be moved and we keep code that is specific to XPU

ZhiweiYan-96 replied:
hi @malfet, Great thanks for your comments. I have removed these small utils and make them inline in xpu files.

malfet reviewed (approved):
It only affects XPU so it's probably fine, but it does a lot of unnecessary code duplication which will make an effort of enabling XPU builds by default harder and harder to achieve

EikanWang commented:
Thanks for the comments, @malfet. We are refining the PR.

ZhiweiYan-96 committed:
Update on "Intel GPU oneDNN upstreaming: Conv primitive integration"

# Motivation

This PR is a part of RFC #114848,. This PR would depend on oneDNN compilation in #117098 and basic integration support in #117112.  Some runtime support is needed in #116019.

This PR provides the oneDNN integration code for Convolution and Deconvolution related operators. All aten convolution operators(conv, deconv, and conv-pointwise fusion) will goes into this layer before executing oneDNN primitive.  The integration code is responsible for providing correct memory description for primitive and accompanied with primitive attribute description.

Wit this PR land, we will add Conv related operators accompanied with their registration.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 commented:
@pytorchbot merge

EikanWang commented:
@pytorchbot merge
